cpp/src/sampling/renumber_sampled_edgelist_impl.cuh

```diff
@@ -0,0 +1,723 @@
+      handle.get_stream());
+    rmm::device_uvector<vertex_t> merged_vertices(merged_label_indices.size(), handle.get_stream());
+    auto label_src_triplet_first =
+      thrust::make_zip_iterator((*unique_label_src_pair_label_indices).begin(),
+                                thrust::make_constant_iterator(uint8_t{0}),
+                                unique_label_src_pair_vertices.begin());
+    auto label_dst_triplet_first =
+      thrust::make_zip_iterator((*unique_label_dst_pair_label_indices).begin(),
+                                thrust::make_constant_iterator(uint8_t{1}),
+                                unique_label_dst_pair_vertices.begin());
+    thrust::merge(
+      handle.get_thrust_policy(),
+      label_src_triplet_first,
+      label_src_triplet_first + (*unique_label_src_pair_label_indices).size(),
+      label_dst_triplet_first,
+      label_dst_triplet_first + (*unique_label_dst_pair_label_indices).size(),
+      thrust::make_zip_iterator(
+        merged_label_indices.begin(), thrust::make_discard_iterator(), merged_vertices.begin()));
+
+    return std::make_tuple(std::move(merged_vertices), std::move(merged_label_indices));
+  } else {
+    rmm::device_uvector<vertex_t> output_vertices(unique_label_dst_pair_vertices.size(),
+                                                  handle.get_stream());
+    auto output_last = thrust::set_difference(
+      handle.get_thrust_policy(),
+      unique_label_dst_pair_vertices.begin(),
+      unique_label_dst_pair_vertices.end(),
+      edgelist_hops ? (*sorted_srcs).begin() : unique_label_src_pair_vertices.begin(),
+      edgelist_hops ? (*sorted_srcs).end() : unique_label_src_pair_vertices.end(),
+      output_vertices.begin());
+
+    sorted_srcs = std::nullopt;
+
+    auto num_unique_srcs = unique_label_src_pair_vertices.size();
+    auto renumber_map    = std::move(unique_label_src_pair_vertices);
+    renumber_map.resize(
+      renumber_map.size() + thrust::distance(output_vertices.begin(), output_last),
+      handle.get_stream());
+    thrust::copy(handle.get_thrust_policy(),
+                 output_vertices.begin(),
+                 output_last,
+                 renumber_map.begin() + num_unique_srcs);
+
+    return std::make_tuple(std::move(renumber_map), std::nullopt);
+  }
+}
+
+}  // namespace
+
+template <typename vertex_t, typename label_t>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<size_t>>>
+renumber_sampled_edgelist(
+  raft::handle_t const& handle,
+  rmm::device_uvector<vertex_t>&& edgelist_srcs,
+  std::optional<raft::device_span<int32_t const>> edgelist_hops,
+  rmm::device_uvector<vertex_t>&& edgelist_dsts,
+  std::optional<std::tuple<raft::device_span<label_t const>, raft::device_span<size_t const>>>
+    label_offsets,
+  bool do_expensive_check)
+{
+  using label_index_t = uint32_t;
```

Comment:
```suggestion
  using label_index_t = int32_t;
```
Shouldn't this be int32_t?

Reply:
This can either be uint32_t or int32_t. Note that `label_t` and `label_index_t` are two different things.

Say, if we can have labels

`10, 25, 30, 7`

and label_indices

`0, 1, 2, 3`.

uint32_t here is basically 4 byte version of size_t (as we may not have more than 2^32-1 labels per GPU in the foreseeable future).

Reply:
Ok, sorry for the confusion with `label_t`.  This makes sense.

Reply:
I am getting an error though:
```
RuntimeError: non-success value returned from cugraph_uniform_neighbor_sample: CUGRAPH_UNKNOWN_ERROR cuGraph failure at file=/root/cugraph/cpp/src/sampling/renumber_sampled_edgelist_impl.cuh line=475: Invalid input arguments: current implementation assumes that the number of unique labels is no larger than std::numeric_limits<uint32_t>::max().
```

Reply:
Oh, I know what's going on.  We need to use a flag to turn renumbering on/off.

Reply:
@seunghwak can you add a flag to this function, and then either Chuck or I can properly set it from the value provided to the C API.

Reply:
And we should also enforce that renumbering can only be done if the label list is provided, which I think is the direct cause of the error here.

Reply:
> Oh, I know what's going on. We need to use a flag to turn renumbering on/off.

This function is a renumbering function, so if the flag is turned off, this function should not be called, am I missing something?

Reply:
Right now, the uniform neighbor sampling functor is always attempting to renumber.  We want to specify whether or not the functor should do that.

Reply:
So shouldn't this be handled above this PR? `renumber_sampled_edgelist` doesn't control whether to renumber or not. I assume this should be handled in the sampling function (sampling function should either call this function or not based on the flag).

Reply:
That's right, I was asking if you'd update the sampling function as well

Reply:
But I can just do that, it's only an extra boolean and an additional check

Reply:
The issue is that the way it's written in this PR, the sampling python tests will fail

Reply:
Yeah... this PR does not touch any code related to sampling. Better be addressed in the sampling PR calling this function.

Reply:
Ok, I'll take care of it